README
ep128emu 2.0.11.2-core1
=======================

ep128emu is an open source, portable emulator of the Enterprise 128,
ZX Spectrum 48/128, Amstrad CPC 464/664/6128 and Videoton TVC computers,
written in C++, and supporting Windows and POSIX platforms (32 and 64
bit Windows and Linux, and MacOS X have been tested).
It implements accurate, high quality hardware emulation, however, the
system requirements are higher than that of most other emulators.
Versions named "-core" have improvements from the libretro core version.

Features
========

General
-------

  * graphical user interface using the FLTK library
  * software (FLTK based) or OpenGL video, with resizable emulator
    window, fullscreen mode, brightness, contrast, gamma, hue, and color
    saturation control; additional features in OpenGL mode only: single
    or double buffered (with synchronization to vertical refresh) mode,
    linear texture filtering, resampling video output to the monitor
    refresh rate, and some display effects: motion blur, scanline
    shading, and (if OpenGL 2.0 shaders are available) PAL TV emulation
  * real time audio output uses the PortAudio library (v18 or v19), with
    support for many native audio APIs (MME/DirectSound/WDM-KS/WASAPI on
    Windows, OSS/ALSA/JACK on Linux, and CoreAudio on MacOS X); high
    quality sample rate conversion with low aliasing; volume control,
    two first order highpass filters with configurable cutoff frequency,
    and an optional parametric equalizer can be applied to the audio
    signal
  * recording audio output to a WAV format sound file
  * recording video and sound output to an AVI format video file, with
    768x576 RLE8 or 384x288 uncompressed YV12 video at 24 to 60 frames
    per second, and 48000 Hz stereo 16-bit PCM audio
  * saving screenshots as 768x576 8-bit PNG format files
  * saving and loading snapshots of the state of the emulated machine
  * demo recording (snapshot combined with stream of keyboard and mouse
    events which can be played back with accurate timing)
  * tape emulation with playback, recording, and setting tape position;
    markers can be created for quick positioning to specific tape
    locations (useful for tapes with multiple files); uses custom file
    format which is PCM audio data with 1 to 8 bits of sample resolution
    and variable sample rate, and header including the table of markers;
    there is also limited (read only) support for EPTE format tape
    files, as well as read-write (although without markers) support for
    sound files like WAV, AIFF, etc.
  * GUI tape editor utility for copying Enterprise files from/to
    ep128emu tape images
  * GUI debugger with support for breakpoints/watchpoints, viewing the
    current state of CPU registers and memory paging, displaying memory
    dump and searching for a pattern of bytes, and disassembler with
    support for all documented and undocumented Z80 opcodes.
    A simple monitor is also included, with commands like assemble,
    disassemble (also to file), trace, memory and I/O port dump and
    modify, printing and changing CPU registers, memory compare, copy,
    fill, search, load and save, and more.
    For most operations, addresses can be 16 bit CPU (affected by
    current paging) or 22 bit physical (all ROM and RAM data can be
    accessed, regardless of memory paging) addresses. Watchpoints can
    also be set on I/O ports and physical addresses.
    The debugger supports scripting in the Lua language, to allow for
    advanced uses like breakpoints with custom defined, complex set of
    conditions.
  * configurable keyboard map for the emulated machine; it is also
    possible to use external game controller devices like joysticks and
    gamepads

Enterprise emulation
--------------------

  * instruction based emulation of the Z80 CPU, supports all documented
    and undocumented opcodes, and memory wait states (including
    synchronization with the NICK chip when accessing video memory)
  * RAM size can be set in 16 kilobyte steps in the range 64 to 3712
  * ROM can be loaded from external image files to segments 0 to 7,
    16 to 19 (decimal), 32 to 35, 48 to 51, and 64 to 67
  * using external configuration files, it is also possible to define
    any memory configuration without the above limitations
  * NICK chip emulation, supporting all documented and undocumented
    video modes
  * DAVE emulation, including timers, interrupts, external ports for
    tape and keyboard/joystick, memory paging, and sound output (all
    effects are supported, and the polynomial counters generate the same
    pseudo-random "noise" pattern as on the real machine)
  * tape emulation (see general features for details)
  * WD177x (floppy drive controller) emulation for EXDOS
  * IDE hard disk emulation; supports up to 4 2 GB disks, image files
    can be in raw or VHD format
  * optional extension ROM (epfileio.rom) that implements a FILE: device
    for direct access to files on the host system in a single user
    selectable directory
  * Spectrum emulator card emulation
  * real time clock (at ports 7E, 7F)
  * external 4-channel 8-bit DAC at ports F0 to F3
  * mouse emulation, an EnterMice device with up to 5 buttons and mouse
    wheel is emulated in native mode on column K of the keyboard matrix
  * SD card (SDEXT) emulation, this is an experimental feature based on
    code from LGB's Xep128 emulator
  * SID card emulation at ports 0E (address) and 0F (data) using reSID
    1.0; NOTE: this currently breaks the external DAC emulation when
    active

Spectrum emulation
------------------

  * instruction based emulation of the Z80 CPU, supports all documented
    and undocumented opcodes, and cycle accurate synchronization with
    the ULA chip when accessing video memory and I/O ports
  * RAM size can be 16, 48, or 128 kilobytes
  * ROM can be loaded from external image files
  * ULA, AY-3-8912, keyboard, and Kempston joystick emulation
  * tape emulation (see general features for details)
  * Spectrum tape files in .tap format can be used as tape images for
    read-only hardware level tape emulation, or loaded directly to
    memory if the "Enable virtual file I/O" option is checked in the
    machine configuration
  * it is also possible to load .tzx format tape images, although the
    support for these is not complete yet
  * loading SNA and Z80 format snapshot files created by other emulators
    is supported; however, snapshot saving and demo recording are only
    possible in the native ep128emu format

Amstrad CPC emulation
---------------------

  * instruction based emulation of the Z80 CPU, supports all documented
    and undocumented opcodes, and cycle accurate synchronization with
    the CRTC and gate array when accessing video memory and I/O ports
  * RAM size can be 64, 128, 192, 320, or 576 kilobytes
  * ROM can be loaded from external image files to the lower ROM and
    expansion ROMs 0 to 7
  * 6845 CRTC (currently type 0 only), gate array, AY-3-8912, 8255 PPI
    (modes 1 and 2 are not supported yet), keyboard, and joystick
    emulation
  * tape emulation (see general features for details)
  * CPC tape files in .cdt (same as Spectrum .tzx) format can be used as
    tape images for read-only hardware level tape emulation, although
    the support for this format is not complete yet
  * uPD765 (floppy drive controller) emulation; it supports standard and
    extended .DSK files, and also access to real disks (not very useful
    in practice, since only PC format disks can be used this way), as
    well as limited emulation of the timing of disk rotation, stepping,
    and data transfers. The FORMAT TRACK (0x0D) command is not
    implemented yet, and some copy protected games and .DSK files with
    unusual track formats may not work correctly
  * loading SNA snapshot files (v1 or v2; v3 is loaded, but the extra
    information about the internal state of the hardware is currently
    ignored) created by other emulators is supported; however, snapshot
    saving and demo recording are only possible in the native ep128emu
    format

Videoton TVC emulation
----------------------

  * instruction based emulation of the Z80 CPU, supports all documented
    and undocumented opcodes, and cycle level synchronization with the
    CRTC and other devices when accessing video memory and I/O ports
  * RAM size can be 48, 80 or 128 kilobytes (TVC 32, 64 or 64+)
  * ROM can be loaded from external image files to SYS, CART and EXT
  * 6845 CRTC (currently type 0 only), audio, keyboard and joystick
    emulation
  * tape emulation (see general features for details)
  * WD1773/1793 (floppy drive controller) emulation for VT-DOS
  * optional extension ROM (tvcfileio.rom) that implements a FILE device
    for direct access to files on the host system in a single user
    selectable directory
  * SD card (SDEXT) emulation, this is an experimental feature based on
    code from LGB's Xep128 emulator

Installation
============

Linux
-----

On Linux and other POSIX platforms, the emulator is installed from the
source code, available at the GitHub download page
  https://github.com/istvan-v/ep128emu/releases/
or the most recent state of the code can be downloaded from Git with the
following command:
  git clone https://github.com/istvan-v/ep128emu.git
In addition to the standard development tools (a recent version of the
GNU C/C++ compiler, binutils, etc.), you need the following packages:

  * SCons (http://www.scons.org/)
  * Python interpreter for running SCons
  * FLTK 1.3.x (http://www.fltk.org/software.php?VERSION=1.3.4) or
    1.1.x (http://www.fltk.org/software.php?VERSION=1.1.10)
    NOTES:
      * this library should be compiled with the --enable-threads
        'configure' option - many Linux distributions include binaries
        of FLTK 1.1 built without --enable-threads, so you may need to
        compile it from sources
      * on MacOS X, FLTK 1.1.7 needs to be patched with the included
        fltk-1.1.7-MacOSX.patch file
  * PortAudio (http://www.portaudio.com/download.html), version 18 and
    19 are supported, but v19 is recommended
  * libsndfile (http://www.mega-nerd.com/libsndfile/#Download)

The following packages are optional:

  * OpenGL for improved video display and effects
  * SDL (http://www.libsdl.org/) 1.2 for joystick input; NOTE: on Linux,
    versions 1.2.10 and newer do not work if they are statically linked
    and were built with video support (--enable-video)
  * Lua (http://www.lua.org/ or http://www.luajit.org/) for scripting in
    the debugger; version 5.1 or newer is recommended, but 5.0 is also
    supported
  * libcurl (https://curl.haxx.se/) for ROM download support in
    epmakecfg

Once these are installed, you can edit the file 'SConstruct' in the top
level source directory for setting compiler flags etc., and run the
command 'scons -j 2' for building the emulator. The following options
can be passed to scons:

  -j N
      Maximum number of parallel jobs to be run by scons. It is
      recommended to set N to the number of logical cores of your CPU
  win64=1
      Cross-compile Windows x86_64 binaries using Wine
      (http://www.winehq.com/) and MinGW; complete pacakges of MinGW for
      building the emulator can be downloaded from:
        https://www.dropbox.com/s/ex0zxz7e6bmcb9r/mingw_w64-x64.7z?dl=1
        https://www.dropbox.com/s/a336x3bs1p0gv3t/mingw_w64-x86.7z?dl=1
  win32=1

NEWS
Changes in version 2.0.11.2-core1
---------------------------

  * fixed TVC compilation without SDEXT emulation
  * backend support for external joysticks 3..6 (no GUI representation)
  * speaker on/off support
  * support of TAP files created by TAPir (read only)
  * crude internal WAV handler for special formats like 1-bit WAV for TVC
  * include ROMs in source (untested here)
  * fixed byte order detection on aarch64 (arm64) platforms
  * other code changes without functional effect, to be able to keep the 
    code similar to the libretro core version

Changes in version 2.0.11.2
---------------------------

  * increased Enterprise maximum sound clock frequency to 1.25 MHz for
    accurate emulation of machines with 10 MHz CPU
  * fixed FILE: bug in re-opening read only files
  * fixed keys "sticking" because of buggy FLTK file chooser on Linux
  * fixed Z80 clock stretching on debug writes to NICK ports
  * experimental support for MIDI port emulation using PortMidi
  * fixed formatting floppy disks with 11 sectors per track
  * fixed floppy drive ready signal in EXDOS
  * added support for comments (beginning with ; or #) in the breakpoint
    list, and reading breakpoints from SjASM export files

Changes in version 2.0.11.1
---------------------------

  * fixed error on Windows with non-ASCII characters in the user name

Changes in version 2.0.11
-------------------------

  * implemented Videoton TVC emulation
  * the epimgconv utility has a graphical user interface now, and it can
    also convert images to TVC format
  * various improvements and new features in the epcompress utility
  * SID card emulation using reSID 1.0
  * the debugger window has been made resizeable within a limited range
  * improved support for Unicode characters in file names on Windows
  * unformatted or non-FAT floppy disk images default to 80/2/9 geometry
    if the file size is 737280 bytes
  * fixed bug in re-opening files selected with a file chooser dialog
  * fixed FILE device bug in EXOS 10 with files larger than 64 KB
  * some other bug fixes

Changes in version 2.0.10
-------------------------

  * implemented mouse (EnterMice) support in the Enterprise emulation
  * when loading a snapshot from the command line, the machine type is
    autodetected from the file if not specified
  * window icons have been implemented on non-Windows platforms if the
    emulator is built with FLTK 1.3.3 or newer
  * on the Enterprise and CPC, the external joysticks can now have a
    second and third fire button, configured as keys 0x75-0x76 (EXT1)
    and 0x7D-0x7E (EXT2)
  * more accurate emulation of NICK port reading and floating data bus
  * improved timing accuracy of NICK and DAVE port I/O (this may break
    compatibility with some demo files recorded with older versions)
  * implemented EXOS function 10 (set and read channel status) in the
    FILE: device and epfileio.rom
  * screenshots are saved in PNG format instead of BMP
  * new Lua functions: writeROM(), writeWordROM(), and loadROMSegment()
  * building with Lua 5.2 or 5.3 is now supported
  * added support for building 64-bit Windows binaries
  * OpenGL has been made an optional dependency at compile time
  * epimgconv and other utilities are included with the emulator package
  * snapshot and demo files can be compressed in epcompress format
  * experimental SD card emulation using code from LGB's Xep128,
    included only if enabled at compile time
  * various bug fixes and minor optimizations

Changes in version 2.0.9.1
--------------------------

  * some debugger changes related to ignore breakpoints
  * it is possible to delete single breakpoints in Lua scripts by using
    a negative priority value
  * minor 6845 CRTC emulation bug fix
  * fixed bugs in closing the debugger window with the Esc key

Changes in version 2.0.9
------------------------

  * implemented CPC floppy drive controller (uPD765) emulation:
    - it supports standard and extended .DSK image files, as well as
      using a real disk in the PC floppy drive (the latter is not very
      useful in practice, since it is limited to standard PC format DD
      (720K) disks, which are not supported by AMSDOS)
    - 4 drives can be emulated (although AMSDOS will only use 2)
    - there is (not perfect) emulation of the timing of disk rotation,
      stepping, head (un)load, and data transfers; however, this may
      not work if the image file includes a track format that is not
      possible on a real DD disk (~6250 bytes per track)
    - the FORMAT TRACK (0x0D) command is unimplemented - it will always
      report a write protect error
    - the floppy drive emulation has not been tested extensively yet,
      so there may be problems with unusual disk formats, rarely used
      FDC commands and parameters, and some copy protections
    - reading the I/O ports 80H-9FH in the debugger will return some
      debugging information about the FDC emulation (see the README file
      for details)
    - changes to the installer/packaging: new machine configuration
      files have been added that include AMSDOS (the default is now
      CPC 6128 with AMSDOS), the disk.zip package includes an empty CPC
      disk image, and on Windows .DSK files can be associated with the
      emulator
  * the CPC video emulation now supports half-character resolution
    scrolling by changing the horizontal sync width
  * improved 6845 CRTC emulation accuracy, fixing problems in some CPC
    games
  * it is possible to disable software control of the tape motor (this
    is useful mainly to work around some buggy Enterprise software)
  * when using the PC floppy drive and a real disk for WD177x emulation,

src/tape.hpp

// ep128emu -- portable Enterprise 128 emulator
// Copyright (C) 2003-2010 Istvan Varga <istvanv@users.sourceforge.net>
// http://sourceforge.net/projects/ep128emu/
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#ifndef EP128EMU_TAPE_HPP
#define EP128EMU_TAPE_HPP

#include "ep128emu.hpp"

#include <vector>
#ifndef EXCLUDE_SOUND_LIBS
#include <sndfile.h>
#endif // EXCLUDE_SOUND_LIBS

namespace Ep128Emu {

  class Tape {
   protected:
    long      sampleRate;       // defaults to 24000
    int       fileBitsPerSample;
    int       requestedBitsPerSample;
    bool      isReadOnly;
    bool      isPlaybackOn;
    bool      isRecordOn;
    bool      isMotorOn;
    size_t    tapeLength;       // tape length (in samples)
    size_t    tapePosition;     // current read/write position (in samples)
    int       inputState;
    int       outputState;
    Tape(int bitsPerSample = 1);
   public:
    virtual ~Tape();
    /*!
     * Get sample rate of tape emulation.
     */
    inline long getSampleRate() const
    {
      return sampleRate;
    }
    /*!
     * Get number of bits (1, 2, 4, or 8) per sample in tape file.
     */
    inline int getSampleSize() const
    {
      return fileBitsPerSample;
    }
    /*!
     * Returns true if the tape file is opened in read-only mode.
     */
    inline bool getIsReadOnly() const
    {
      return isReadOnly;
    }
   protected:
    virtual void runOneSample_();
   public:
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
    inline void runOneSample()
    {
      if (isPlaybackOn && isMotorOn)
        runOneSample_();
    }
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Returns true if the motor is currently on.
     */
    inline bool getIsMotorOn() const
    {
      return isMotorOn;
    }
    /*!
     * Start playback.
     * (note: the motor should also be turned on to actually play)
     */
    inline void play()
    {
      isPlaybackOn = true;
      isRecordOn = false;
    }
    /*!
     * Start recording; if the tape file is read-only, this is equivalent
     * to calling play().
     * (note: the motor should also be turned on to actually record)
     */
    inline void record()
    {
      isPlaybackOn = true;
      isRecordOn = !isReadOnly;
    }
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Set input signal for recording.
     */
    inline void setInputSignal(int newState)
    {
      inputState = newState;
    }
    /*!
     * Get output signal.
     */
    inline int getOutputSignal() const
    {
      return outputState;
    }
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Returns the current tape position in seconds.
     */
    inline double getPosition() const
    {
      return (double(uint32_t(tapePosition)) / double(sampleRate));
    }
    /*!
     * Returns the current tape length in seconds.
     */
    inline double getLength() const
    {
      return (double(uint32_t(tapeLength)) / double(sampleRate));
    }
    /*!
     * Returns true if the current position is at the end of the tape;
     * should be used only when reading a tape file.
     */
    inline bool getIsEndOfTape() const
    {
      return (tapePosition >= tapeLength);
    }
    /*!
     * Seek forward (if isForward = true) or backward (if isForward = false)
     * to the nearest cue point, or by 't' seconds if no cue point is found.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Create a new cue point at the current tape position.
     * Has no effect if the file does not have a cue point table, or it
     * is read-only.
     */
    virtual void addCuePoint();
    /*!
     * Delete the cue point nearest to the current tape position.
     * Has no effect if the file is read-only.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Delete all cue points. Has no effect if the file is read-only.
     */
    virtual void deleteAllCuePoints();
  };

  class Tape_Ep128Emu : public Tape {
   private:
    std::FILE *f;               // tape image file
    uint8_t   *buf;             // 4096 bytes
    uint32_t  *fileHeader;      // table with a length of 1024, contains:
                                //      0, 1: magic number (0x0275CD72,
                                //            0x1C445126)
                                //         2: number of bits per sample in
                                //            file (1, 2, 4, or 8)
                                //         3: sample rate in Hz (10000 to
                                //            120000)
                                // 4 to 1022: the sample positions where cue
                                //            points are defined, in sorted
                                //            order, padded with 0xFFFFFFFF
                                //            values
                                //      1023: must be 0xFFFFFFFF
    size_t    cuePointCnt;
    bool      isBufferDirty;    // true if 'buf' has been changed,
                                // and not written to file yet
    bool      usingNewFormat;
    // ----------------
    void seek_(size_t pos_);
    bool findCuePoint_(size_t& ndx_, size_t pos_);
    void packSamples_();
    bool writeBuffer_();
    bool readBuffer_();
    void unpackSamples_();
    bool writeHeader_();
    void flushBuffer_();
   public:
    /*!
     * Open tape file 'fileName'. If the file does not exist yet, it may be
     * created (depending on the 'mode' parameter) with the specified sample
     * rate and bits per sample. Otherwise, 'sampleRate_' is ignored, and
     * samples are converted according to 'bitsPerSample'.
     * 'mode' can be one of the following values:
     *   0: open tape file read-write if possible, or create a new file if
     *      it does not exist
     *   1: open tape file read-write if possible, and fail if it does not
     *      exist
     *   2: open an existing tape file read-only
     *   3: create new tape file in read-write mode; if the file already
     *      exists, it is truncated and a new header is written
     */
    Tape_Ep128Emu(const char *fileName, int mode = 0,
                  long sampleRate_ = 24000L, int bitsPerSample = 1);
    virtual ~Tape_Ep128Emu();
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
   protected:
    virtual void runOneSample_();
   public:
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Seek forward (if isForward = true) or backward (if isForward = false)
     * to the nearest cue point, or by 't' seconds if no cue point is found.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Create a new cue point at the current tape position.
     * Has no effect if the file does not have a cue point table, or it
     * is read-only.
     */
    virtual void addCuePoint();
    /*!
     * Delete the cue point nearest to the current tape position.
     * Has no effect if the file is read-only.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Delete all cue points. Has no effect if the file is read-only.
     */
    virtual void deleteAllCuePoints();
  };

  class Tape_WAV : public Tape {
   private:
    std::FILE *f;               // wave file
    uint8_t   *buf;             // 4096 bytes
    uint32_t  *fileHeader;      // 44 byte long header:
/*
0         4   ChunkID          Contains the letters "RIFF" in ASCII form
4         4   ChunkSize        ignored here
8         4   Format           Contains the letters "WAVE"
12        4   Subchunk1ID      Contains the letters "fmt "
16        4   Subchunk1Size    16 for PCM.
20        2   AudioFormat      PCM = 1 (i.e. Linear quantization)
22        2   NumChannels      Mono = 1, Stereo = 2, etc.
24        4   SampleRate       8000, 44100, etc.
28        4   ByteRate         ignored here
32        2   BlockAlign       ignored here
34        2   BitsPerSample    8 bits = 8, 16 bits = 16, etc.
36        4   Subchunk2ID      Contains the letters "data"
40        4   Subchunk2Size    ignored here
44        *   Data             The actual sound data.
*/
    bool      usingNewFormat;
    // ----------------
    bool readBuffer_();
    void unpackSamples_();
   public:
    /*!
     * Open tape file 'fileName'.
     */
    Tape_WAV(const char *fileName, int bitsPerSample = 1);
    virtual ~Tape_WAV();
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
   protected:
    virtual void runOneSample_();
   public:
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Not supported for this format.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Not supported for this format.
     */
    virtual void addCuePoint();
    /*!
     * Not supported for this format.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Not supported for this format.
     */
    virtual void deleteAllCuePoints();
  };



  class Tape_EPTE : public Tape {
   private:
    std::FILE *f;
    size_t    bytesRemaining;
    bool      endOfTape;
    uint8_t   shiftReg;
    uint8_t   bitsRemaining;
    uint8_t   halfPeriodSamples;
    size_t    samplesRemaining;
    size_t    leaderSampleCnt;
    size_t    chunkBytesRemaining;
    size_t    chunkCnt;
    uint32_t  chunkArray[128]; // for EPTE/TAPir chunk pointers stored in header
   public:
    /*!
     * Open EPTE format tape file 'fileName' read-only.
     */
    Tape_EPTE(const char *fileName, int bitsPerSample = 1);
    virtual ~Tape_EPTE();
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
   protected:
    virtual void runOneSample_();
   public:
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Seek forward (if isForward = true) or backward (if isForward = false)
     * to the nearest cue point, or by 't' seconds if no cue point is found.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Create a new cue point at the current tape position.
     * Has no effect if the file does not have a cue point table, or it
     * is read-only.
     */
    virtual void addCuePoint();
    /*!
     * Delete the cue point nearest to the current tape position.
     * Has no effect if the file is read-only.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Delete all cue points. Has no effect if the file is read-only.
     */
    virtual void deleteAllCuePoints();
  };

  class Tape_TZX : public Tape {
   private:
    std::FILE *f;
    uint8_t   currentBlockType;
    uint8_t   currentMode;
    bool      endOfTape;
    uint8_t   shiftReg;
    uint32_t  pulseTimer;
    uint32_t  pulseLength;
    uint32_t  pulseCnt;
    uint16_t  pilotPulseLength;
    uint16_t  syncPulseLength1;
    uint16_t  syncPulseLength2;
    uint16_t  bit0PulseLength;
    uint16_t  bit1PulseLength;
    uint8_t   bit0PulseCnt;
    uint8_t   bit1PulseCnt;
    uint16_t  pilotPulseCnt;
    uint8_t   lastByteBits;
    uint8_t   pulseSequencePulsesLeft;
    uint32_t  pauseLength;
    uint32_t  clockFrequency;
    uint32_t  dataBlockBytesLeft;
    uint32_t  directRecordingSampleRate;
    uint32_t  directRecordingTimer;
    uint32_t  loopFilePos;
    size_t    loopStartTime;
    uint16_t  loopRepeatCnt;
    bool      isTAPFile;
   public:
    /*!
     * Open TZX or Spectrum TAP format tape file 'fileName' read-only.
     */
    Tape_TZX(const char *fileName, int bitsPerSample = 1);
    virtual ~Tape_TZX();
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
   protected:
    void tapeReset();
    bool readByte(uint8_t& n);
    bool readUInt16(uint16_t& n);
    bool readUInt24(uint32_t& n);
    bool readUInt32(uint32_t& n);
    uint32_t convertPulseLength(uint32_t n, uint32_t clockFreq_ = 0U);
    void setPauseMode(uint32_t pauseLength_ = 0U);
    void readNextTZXBlock();
    void directRecordingNextBit();
    void dataBlockNextBit();
    virtual void runOneSample_();
   public:
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Seek forward (if isForward = true) or backward (if isForward = false)
     * to the nearest cue point, or by 't' seconds if no cue point is found.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Create a new cue point at the current tape position.
     * Has no effect if the file does not have a cue point table, or it
     * is read-only.
     */
    virtual void addCuePoint();
    /*!
     * Delete the cue point nearest to the current tape position.
     * Has no effect if the file is read-only.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Delete all cue points. Has no effect if the file is read-only.
     */
    virtual void deleteAllCuePoints();
  };

#ifndef EXCLUDE_SOUND_LIBS
  class Tape_SoundFile : public Tape {
   public:
    class TapeFilter {
     private:
      std::vector<float> irFFT;
      std::vector<float> fftBuf;
      std::vector<float> outBuf;
      size_t  sampleCnt;
     public:
      TapeFilter(size_t irSamples = 1024);
      virtual ~TapeFilter();
      void setFilterParameters(float sampleRate, float minFreq, float maxFreq);
      float processSample(float inputSignal);
      static void fft(float *buf, size_t n, bool isInverse);
    };
   private:
    SNDFILE     *sf;            // tape image file
    std::vector<short>  buf;    // 1024 interleaved sample frames
    int         nChannels;
    int         requestedChannel;
    bool        enableFIRFilter;
    bool        isBufferDirty;  // true if 'buf' has been changed,
                                // and not written to file yet
    TapeFilter  firFilter;
    // ----------------
    void seek_(size_t pos_);
    bool writeBuffer_();
    void flushBuffer_();
   public:
    /*!
     * Open tape file 'fileName'.
     * 'mode' can be one of the following values:
     *   0, 1: open tape file read-write if possible, and fail if it does not
     *         exist
     *   2:    open an existing tape file read-only
     */
    Tape_SoundFile(const char *fileName, int mode = 0, int bitsPerSample = 1);
    virtual ~Tape_SoundFile();
    /*!
     * Run tape emulation for a period of 1.0 / getSampleRate() seconds.
     */
   protected:
    virtual void runOneSample_();
   public:
    /*!
     * Turn motor on (newState = true) or off (newState = false).
     */
    virtual void setIsMotorOn(bool newState);
    /*!
     * Stop playback and recording.
     */
    virtual void stop();
    /*!
     * Seek to the specified time (in seconds).
     */
    virtual void seek(double t);
    /*!
     * Seek forward (if isForward = true) or backward (if isForward = false)
     * to the nearest cue point, or by 't' seconds if no cue point is found.
     */
    virtual void seekToCuePoint(bool isForward = true, double t = 10.0);
    /*!
     * Create a new cue point at the current tape position.
     * Has no effect if the file does not have a cue point table, or it
     * is read-only.
     */
    virtual void addCuePoint();
    /*!
     * Delete the cue point nearest to the current tape position.
     * Has no effect if the file is read-only.
     */
    virtual void deleteNearestCuePoint();
    /*!
     * Delete all cue points. Has no effect if the file is read-only.
     */
    virtual void deleteAllCuePoints();
    /*!
     * Set parameters for sound file reading.
     */
    void setParameters(int requestedChannel_, bool enableFIRFilter_,
                       float filterMinFreq_, float filterMaxFreq_);
  };
#endif // EXCLUDE_SOUND_LIBS

  /*!
   * Open tape file 'fileName'. If the file does not exist yet, it may be
   * created (depending on the 'mode' parameter) with the specified sample
   * rate and bits per sample. Otherwise, 'sampleRate_' is ignored, and
   * samples are converted according to 'bitsPerSample'.
   * 'mode' can be one of the following values:
   *   0: open tape file read-write if possible, or create a new file if
   *      it does not exist
   *   1: open tape file read-write if possible, and fail if it does not
   *      exist
   *   2: open an existing tape file read-only
   *   3: create new tape file in read-write mode; if the file already
   *      exists, it is truncated and a new header is written
   */
  Tape *openTapeFile(const char *fileName, int mode = 0,
                     long sampleRate_ = 24000L, int bitsPerSample = 1);

}       // namespace Ep128Emu

#endif  // EP128EMU_TAPE_HPP

